### ISSUE TITLE
[IDEA] slugify operator - transliterate period (.)

### ISSUE THREAD

**Issue by twMat** [TiddlyWiki/TiddlyWiki5#4660]
[Slugify was just implemented and here Mat goes right away...]

When using the slugify operator to construct style class names from tiddler titles, any periods `.` in the title must be transliterated for the name to be allowed as a class name.

I wish for slugs to be usable in the three use cases:
* url slugs
* ~~field names~~ (as of 5.2.0 this is not a problem)
* class names

For the latter, periods would suitably be transliterated into underscores `_`.

(For class names, underscore `_` is preferable over hyphen `-` because of [w3.org states](https://www.w3.org/TR/CSS21/syndata.html#characters) that classnames cannot start with "two hyphens or a hyphen followed by a digit")

A way around the problem is to add `split[.]join[]` before the `slugify[]` but it would be preferable with a real solution.

**Comment by Jermolene** [comment]
Hi @twMat that's helpful. Conversely, for URL slugs it's quite handy to retain periods because it means that tiddlers like "myfile.html" and "myfile.txt" will end up with the expected periods in the title. So, I think we're going to have to add some switches for controlling exactly which transfomations are applied.

**Comment by twMat** [comment]
Apropos: It would be desirable if the transliterations were more transparent to the user. Maybe (I didn't think this through) they could be in a separate (shadow) tiddler so that they could be user modified.

**Comment by saqimtiaz** [comment]
There are two other options that should be considered apart from extending the `slugify` operator:

- we now have a quite powerful `search-replace` operator that could be used for very specific edge cases.
- For things like CSS class names, we extend the format operator, for example: `[<tiddler-title>format:classname[]]`

**Comment by twMat** [comment]
@saqimtiaz - interesting ideas but would they put the burden on the user to construct the transliterations? IMO that would be unfortunate.

Somewhat OT but maybe the slugify operator should be deprecated/replaced with a suffix to the `format` operator, i.e `format:slugify`? ... but then the OP here is still an issue.

Regarding you proposal for `format:classname` - how would this differ from a `slugify:classname`? I mean, for either to be meaningful, they'd have both have to do a full slugification (+ the "classnamification")?

**Comment by saqimtiaz** [comment]
@twMat 

Firstly the intent with my previous post isn't to necessarily supplant the ideas presented in this issue. Rather I am pointing out that we do now have more options that can be used until this issue is considered resolved.

Furthermore, any particularly complex transliteration rules could be handled by `search-replace` instead of adding too much complexity to the slugify operator. 

Lastly I feel `format:classname[]` is more intuitive given that we now have a `format` operator, rather than `slugify:classname`. It is a smaller mental leap in my opinion, to realize that you need to format a title for class names and therefore think of the `format` operator, rather than make the connection to `slugify` with a special suffix.

I think that there is merit to the idea of replacing `slugify[]` with `format:slugify[]`

Thoughts @Jermolene ?

**Comment by twMat** [comment]
>I think that there is merit to the idea of replacing `slugify[]` with `format:slugify[]`

The OP lists three use cases. Thus I'd like to see:

* url slugs  ---------> `format:slugify[]` (...or should probably be `format:slug[]`)
* field names -----> `format:fieldname[]`
* class names -----> `format:classname[]`

**Comment by saqimtiaz** [comment]
@twMat note that just like `slugify`, the output might not be unique for any of them. So for instance you cannot expect a unique classname from a tiddler title.

If we want to deprecate `slugify` it needs to be before 5.1.23 is released.

**Comment by Jermolene** [comment]
I agree that it would be good to add more options to the 'slugify' filter, but I don't see the case for deprecating it in favour of 'format:slugify'. Conversely, if there is an argument for doing so, why wouldn't we do it for 'uppercase', 'lowercase' etc.

I think the 'format' operator is there to match users expectations from using tools like Excel and some programming languages. It's a usage that particularly applies to dates, times and currencies. I'd suggest that the scope of it is to convert to *human readable* formats; if we treat is as a generic text transformation primitive it'd cover everything.
